A51-custom-backend-metrics.md

```diff
@@ -279,67 +303,86 @@ public final class CallMetricRecorder {
+  public CallMetricRecorder recordQpsMetric(double value);
+
   // Returns the call metric recorder attached to the current call.
   public static CallMetricRecorder getCurrent();
 }
 ```
 
-At the client side, Java will provide a utility function `OrcaClientStreamTracerFactory`. 
-Users can include it in the picker result in their custom LB policy’s 
-`LoadBalancer.SubchannelPicker` to let gRPC push notifications to the application.
-It requires users to implement `OrcaPerRequestReportListener` in order to receive the metrics report callback.
-To avoid double-parsing trailers, the utility will use a CallOption in StreamInfo.
-If the call option is not already present, the utility will create an object to parse the trailer
-and notify listeners. The call option is then added and references that created object. If the call
-option is already present, the listener will be added to the list of listeners to be notified.
-
-```Java
-class CustomPicker extends SubchannelPicker {
-  public PickResult pickSubchannel(PickSubchannelArgs args) {
-  Subchannel subchannel = … // WRR picking logic
-  return PickResult.withSubchannel(
-    subchannel,
-    OrcaPerRequestUtil.newOrcaClientStreamTracerFactory(listener));
-  }
-}
-```
-
-#### Out-of-Band Reporting 
-At the server side, Java implemented the OOB service `OrcaServiceImpl`: users can optionally provide the minimum
-reporting interval configuration when creating the object. Then they can update utilization metrics
-data using the exposed APIs as follows. All the methods are thread safe.
+#### Per-Server Recording on Server
+Java will provide a per-server metrics recorder on the server as follows. The user
+creates an instance and records metrics to it. All the methods are thread safe.
+This is optional for per-request load reporting and required for OOB load reporting.
 
 ```java
 public class MetricRecorder {
   // Update the metrics value corresponding to the specified key.
   public void putUtilizationMetric(String key, double value);
 
   // Replace the whole metrics data using the specified map.
-  public void setAllUtilizationMetrics(Map<String, Double> metrics);
+  public void putAllUtilizationMetrics(Map<String, Double> metrics);
 
   // Remove the metrics data entry corresponding to the specified key.
   public void clearUtilizationMetric(String key);
 
   // Update the CPU utilization metrics data.
-  public void setCPUUtilizationMetric(double value);
+  public void putCPUUtilizationMetric(double value);
 
   // Clear the CPU utilization metrics data.
   public void clearCPUUtilizationMetric();
 
   // Update the memory utilization metrics data.
-  public void setMemoryUtilizationMetric(double value);
+  public void putMemoryUtilizationMetric(double value);
 
   // Clear the memory utilization metrics data.
   public void clearMemoryUtilizationMetric();
+
+  // Update the queries-per-second metrics data.
+  public void putQpsMetric(double value);
```

Comment:
@YifeiZhuang, this has "metric" in the method name. The API you added in https://github.com/grpc/grpc-java/pull/9866 was `setQps()`. Should it be `setQpsMetric()`?

Reply:
yes,  `setQpsMetric()` is better. I will make that change in java.